src/udp/socket.rs

```diff
@@ -958,49 +963,117 @@ mod test {
 
     use util;
 
-    mod send_final_ack_acks {
-        use super::*;
-
-        fn localhost_addr(sock: &UdpSocket) -> SocketAddr {
-            let addr = unwrap!(sock.local_addr());
-            SocketAddr::new(ip!("127.0.0.1"), addr.port())
-        }
-
-        #[test]
-        fn it_sends_5_ack_ack_messages() {
-            let mut core = unwrap!(Core::new());
-            let handle = core.handle();
+    fn localhost_addr(sock: &UdpSocket) -> SocketAddr {
+        let addr = unwrap!(sock.local_addr());
+        SocketAddr::new(ip!("127.0.0.1"), addr.port())
+    }
 
-            let recv_sock = unwrap!(UdpSocket::bind(&addr!("0.0.0.0:0"), &handle));
-            let sock = unwrap!(UdpSocket::bind(&addr!("0.0.0.0:0"), &handle));
+    #[test]
+    fn hole_puncher_sends_5_ack_ack_messages() {
```

Comment:
This is quite complex test. If you extracted this block https://github.com/ustulation/p2p/pull/53/files#diff-fcace3d5b60021fa28170b42f248c622R666 - line 666 - 706, I believe that given specific phase (AckAck) you were able to test this behavior (5 acks acks in the end) easier :)

Reply:
True.. though this test tests slightly more behaviour aswell.

Reply:
In such case I like to have separate tests for different scenarios: easier to read and modify, when something fails, error is better localized, when behavior removed, easier to remove tests as well. See: http://www.betterspecs.org/#single